File: php/features.php
<?php
declare(strict_types=1);

// Gitlab SDK feature factory

require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/feature/DebugFeature.php';
require_once __DIR__ . '/feature/IdempotencyFeature.php';
require_once __DIR__ . '/feature/MetricsFeature.php';
require_once __DIR__ . '/feature/PagingFeature.php';
require_once __DIR__ . '/feature/RatelimitFeature.php';
require_once __DIR__ . '/feature/RetryFeature.php';
require_once __DIR__ . '/feature/TestFeature.php';
require_once __DIR__ . '/feature/TimeoutFeature.php';


class GitlabFeatures
{
    public static function make_feature(string $name)
    {
        switch ($name) {
            case "base":
                return new GitlabBaseFeature();
            case "debug":
                return new GitlabDebugFeature();
            case "idempotency":
                return new GitlabIdempotencyFeature();
            case "metrics":
                return new GitlabMetricsFeature();
            case "paging":
                return new GitlabPagingFeature();
            case "ratelimit":
                return new GitlabRatelimitFeature();
            case "retry":
                return new GitlabRetryFeature();
            case "test":
                return new GitlabTestFeature();
            case "timeout":
                return new GitlabTimeoutFeature();
            default:
                return new GitlabBaseFeature();
        }
    }

    /**
     * Does a generated feature class back this name? False for a name only
     * an options extend instance can supply (the station adopt path) - the
     * constructor uses this to skip make_feature for such names instead of
     * adding a stray BaseFeature.
     */
    public static function has_feature(string $name): bool
    {
        switch ($name) {
            case "base":
            case "debug":
            case "idempotency":
            case "metrics":
            case "paging":
            case "ratelimit":
            case "retry":
            case "test":
            case "timeout":
                return true;
            default:
                return false;
        }
    }
}

File: rb/features.rb
# Gitlab SDK feature factory

require_relative 'feature/base_feature'
require_relative 'feature/debug_feature'
require_relative 'feature/idempotency_feature'
require_relative 'feature/metrics_feature'
require_relative 'feature/paging_feature'
require_relative 'feature/ratelimit_feature'
require_relative 'feature/retry_feature'
require_relative 'feature/test_feature'
require_relative 'feature/timeout_feature'


module GitlabFeatures
  def self.make_feature(name)
    case name
    when "base"
      GitlabBaseFeature.new
    when "debug"
      GitlabDebugFeature.new
    when "idempotency"
      GitlabIdempotencyFeature.new
    when "metrics"
      GitlabMetricsFeature.new
    when "paging"
      GitlabPagingFeature.new
    when "ratelimit"
      GitlabRatelimitFeature.new
    when "retry"
      GitlabRetryFeature.new
    when "test"
      GitlabTestFeature.new
    when "timeout"
      GitlabTimeoutFeature.new
    else
      GitlabBaseFeature.new
    end
  end
end
